validator_worker/src/sentry_interface.rs

```diff
@@ -1,143 +1,265 @@
-use primitives::{Channel, ValidatorId};
-use primitives::adapter::{Adapter}
+use crate::error::ValidatorWorkerError;
 use futures::compat::Future01CompatExt;
 use futures::future::{ok, try_join_all, FutureExt, TryFutureExt};
 use futures::Future;
 use futures_legacy::Future as LegacyFuture;
+use primitives::adapter::Adapter;
+use primitives::sentry::{
+    ChannelAllResponse, EventAggregateResponse, LastApprovedResponse, SuccessResponse,
+    ValidatorMessageResponse,
+};
+use primitives::validator::MessageTypes;
+use primitives::{Channel, Config, ValidatorDesc};
+use reqwest::header::AUTHORIZATION;
 use reqwest::r#async::{Client, Response};
-use reqwest::Error;
 use serde::Deserialize;
+use std::error::Error;
 use std::iter::once;
+use std::pin::Pin;
+use std::time::Duration;
 
-
-pub fn all_channels(
-    &self,
-    validator: Option<&ValidatorId>,
-) -> impl Future<Output = Result<Vec<Channel>, Error>> {
-    let validator = validator.cloned();
-    // call Sentry and fetch first page, where validator = identity
-    let first_page = self.clone().fetch_page(1, validator.clone());
-
-    let handle = self.clone();
-    first_page
-        .and_then(move |response| {
-            let first_page_future = ok(response.channels).boxed();
-
-            if response.total_pages < 2 {
-                // if there is only 1 page, return the results
-                first_page_future
-            } else {
-                // call Sentry again for the rest of tha pages
-                let futures = (2..=response.total_pages)
-                    .map(|page| {
-                        handle
-                            .clone()
-                            .fetch_page(page, validator.clone())
-                            .map(|response_result| {
-                                response_result.and_then(|response| Ok(response.channels))
-                            })
-                            .boxed()
-                    })
-                    .chain(once(first_page_future));
-
-                try_join_all(futures)
-                    .map(|result_all| {
-                        result_all
-                            .and_then(|all| Ok(all.into_iter().flatten().collect::<Vec<_>>()))
-                    })
-                    .boxed()
-            }
-        })
-        .boxed()
-}
-
-async fn fetch_page(
-    self,
-    page: u64,
-    validator: Option<ValidatorId>,
-) -> Result<ChannelAllResponse, reqwest::Error> {
-    let mut query = vec![format!("page={}", page)];
-
-    if let Some(validator) = validator {
-        query.push(format!("validator={}", validator));
-    }
-
-    let future = self
-        .client
-        .get(format!("{}/channel/list?{}", self.sentry_url, query.join("&")).as_str())
-        .send()
-        .and_then(|mut res: Response| res.json::<ChannelAllResponse>());
-
-    await!(future.compat())
-}
-
-
-#[derive(Clone)]
-pub(crate) struct SentryApi {
+#[derive(Clone, Debug)]
+pub struct SentryApi<T: Adapter> {
```

Comment:
Is this acting like a service or is it storing data? I am confused.
If you want it to be a service, then it should be stateless, i.e. no `validator` and `channel`. Since the `adapter`, `sentry_url` and etc. can be used for any `validator`+`channel` and then don't reveal state.

Reply:
No that is incorrect, the sentry_url could be different from the validator_url. I have made a change to the name to make it distinct. A validator_url is specified in the channel spec while the sentry_url is passed a flag during startup

Reply:
Also a Sentry is channel specific hence adding the `channel` property

Reply:
@elpiel it's not storing data, but it's constructed with sentry_url and channel as context